lib/core/context.c

```diff
@@ -190,3 +190,30 @@ void h2o_context_update_timestamp_string_cache(h2o_context_t *ctx)
     h2o_time2str_rfc1123(ctx->_timestamp_cache.value->rfc1123, &gmt);
     h2o_time2str_log(ctx->_timestamp_cache.value->log, ctx->_timestamp_cache.tv_at.tv_sec);
 }
+
+void h2o_context_close_idle_connections(h2o_context_t *ctx, size_t max_connections_to_close, uint64_t min_age)
+{
+    if (max_connections_to_close <= 0)
+        return;
+
+    size_t closed = ctx->_conns.num_conns.shutdown;
+
+    if (closed >= max_connections_to_close)
+        return;
+
+    H2O_CONN_LIST_FOREACH(h2o_conn_t * conn, ({&ctx->_conns.idle}), {
+        struct timeval now = h2o_gettimeofday(ctx->loop);
+        if (h2o_timeval_subtract(&conn->connected_at, &now) < (min_age * 1000))
+            continue;
+        ctx->connection_stats.idle_closed++;
+        conn->callbacks->close_idle_connection(conn);
+        closed++;
+        if (closed == max_connections_to_close)
+            return;
+    });
+}
+
+size_t *get_connection_state_counter(h2o_context_t *ctx, h2o_connection_state_t state)
```

Comment:
I really like the idea of abstracting the counters. That said, would it make sense to change this to something like `h2o_set_conn_state(h2o_conn_t *conn, h2o_conn_state_t new_state)`, where:
* the counter is automatically decremented for the old state, 
* incremented for the new state, and
* the state is updated to the given state?

I ask this due to the following reasons:
* We see lots of `--*get_connection_state_counter(...)` followed by `++*get_connection_state_counter(...)`, it's easy to miss `++` after `--`.
* I'm not sure if there's a reason to separate state transition (i.e., assignment to `h2o_conn_t::state`) and counter updates, considering that state is a member of `h2o_conn_t` (which is the superclass of all the conn_t's).


FWIW, `h2o_http2_stream_state` is designed like that.

Reply:
Added some commits, this looks much better and is less error prone. Thanks!